File: app/HDoc/fay/src/Dm/Captcha.hs
-- Copyright 27-Jan-2020 ºDeme
-- GNU General Public License - V3 <http://www.gnu.org/licenses/>

{-# LANGUAGE NamedFieldPuns #-}

--- Capcha widget.

module Dm.Captcha
  ( Init(..)
  , T
  , new
  , wg
  , isZero
  , isOutOfLimit
  , inc
  , reset
  , check
  , newCombination
  ) where

import qualified Data.Var as Var
import Dm.Dom (Element, Property(..), q, q', isChecked)
import Data.Function (fmap)
import qualified Dm.Time as Time
import qualified Dm.Storage as Storage
import qualified Dm.Maybe as Maybe
import qualified Dm.Either as Either
import qualified Dm.Js as Js
import qualified Dm.Rnd as Rnd

--- Init
---  { storeId :: String
---  , counterLimit :: Maybe Int
---  , timeLimit :: Maybe Int
---  , zeroColor :: Maybe String
---  , oneColor :: Maybe String
---  }
---   storeId     : Local Store identifier. Names 'storeId' + "_counter" and
---                 'storeId + "_time' will be used.
---   counterLimit: Number of tries before showing captcha widget. Default 3.
---   timeLimit   : Milliseconds after which captcha counter is set to 0.
---                 Default 900000 (15 minutes)
---   zeroColor   : Color of cells to not mark.
---   oneColor    : color of cells to mark-.
data Init = Init
  { storeId :: String
  , counterLimit :: Maybe Int
  , timeLimit :: Maybe Int
  , zeroColor :: Maybe String
  , oneColor :: Maybe String
  }

--- T
--- Captcha type
data T = T
  { sId :: String
  , cLimit :: Int
  , tLimit :: Int
  , counter :: Var.Ref Int
  , combination :: Var.Ref [Int]

  , wg :: Element
  , zColor :: String
  , oColor :: String
  , checks :: [Element]
  }

data Cmd = Inc | Reset | NewCombination

-- PRIVATE

getCounter :: String -> Fay Int
getCounter sid = do
  now <- Time.now
  tm <- Storage.get (sid ++ "_time")
  ct <- Storage.get (sid ++ "_counter")
  case getTime tm of
    Nothing -> return 0
    Just t -> if Time.df now t > 0 then return 0
              else return $ Maybe.fromMaybe 0 $ getCounter' ct
  where
    getTime tm = tm Maybe.>>= \s ->
                   (Either.toMaybe $ Js.fromStr s Either.>>= Time.fromJs)
    getCounter' ct = ct Maybe.>>= \s ->
                       (Either.toMaybe $ Js.fromStr s Either.>>= Js.ri)

-- CONSTRUCTOR

--- new ini
--- Constructor
new :: Init -> Fay T
new ini = do
  counter <- getCounter (storeId ini) >>= Var.newRef
  combination <- fmap (take 4) (Rnd.shuffle [0..7::Int]) >>= Var.newRef
  wg <- q "div" [] []
  checks <- forM [1..8::Int]
                 (\_ -> q "input" [Att "type" "checkbox"] [] >>= return)
  let r = T
            { sId = storeId ini
            , cLimit = Maybe.fromMaybe 3 $ counterLimit ini
            , tLimit = Maybe.fromMaybe 900000 $ timeLimit ini
            , counter
            , combination

            , wg
            , zColor = Maybe.fromMaybe "#f0f0f0" $ zeroColor ini
            , oColor = Maybe.fromMaybe "#c0c0c0" $ oneColor ini
            , checks
            }
  view r
  return r

-- INFORMATION

--- isZero cp
--- Returns 'True' if captcha counter is zero.
isZero :: T -> Fay Bool
isZero cp = fmap (== 0) $ Var.get (counter cp)

--- isOutOfLimit cp
--- Returns 'True' if captcha counter is greater than its counter limit.
isOutOfLimit :: T -> Fay Bool
isOutOfLimit cp = fmap (>= (cLimit cp)) $ Var.get (counter cp)

--- check cp
--- Returns 'True' if checkboxes are correctly marked.
check :: T -> Fay Bool
check cp = do
  comb <- Var.get (combination cp)
  return $ all (eq comb) $ zip [(0::Int)..] (checks cp)
  where
    eq cb (ix, ck) = any (== ix) cb == isChecked ck

-- VIEW

view :: T -> Fay ()
view model = do
  mapM_ (\ck -> q' ck [Checked False] [] >> return ()) (checks model)
  tds <- mapM (\ck -> q "td" [] [return ck] >>= return) (checks model)
  mapM_ setTd $ zip [(0::Int)..] tds
  let row1 = map return $ take 4 tds
  let row2 = map return $ drop 4 tds
  _ <- q' (wg model) [RemoveAll]
        [ q "table" [Style "border: 1px solid;background-color: #fffff0"]
            [ q "tr" [] row1
            , q "tr" [] row2
            ]
        ]
  return ()
  where
    setTd (ix, td) = do
      comb <- Var.get (combination model)
      let color = case find (ix ==) comb of
                    Nothing -> zColor model
                    _ -> oColor model
      let st = "border: 1px solid;background-color:" ++ color
      q' td [ Style st] [] >> return ()


-- UPDATE

update :: T -> Cmd -> Fay ()
update cp cmd =
  case cmd of
    Inc -> do
      oldCt <- getCounter $ sId cp
      now <- Time.now
      Storage.set (sId cp ++ "_time") $
        Js.toStr $ Time.toJs $ Time.add (tLimit cp) now
      let ct = oldCt + 1
      Storage.set (sId cp ++ "_counter") $ Js.toStr $ Js.wi ct
      Var.set (counter cp) ct
    Reset -> do
      Storage.set (sId cp ++ "_counter") $ Js.toStr $ Js.wi 0
      Var.set (counter cp) 0
    NewCombination -> do
      newComb <- fmap (take 4) (Rnd.shuffle [0..7::Int])
      Var.set (combination cp) newComb
      view cp

--- inc cp
--- Increments captcha counter.
inc :: T -> Fay ()
inc cp = update cp Inc

--- reset cp
--- Sets captcha counter to 0.
reset :: T -> Fay ()
reset cp = update cp Reset

--- newCombination
--- Changes checkboxes combination and modifies view.
newCombination :: T -> Fay()
newCombination cp = update cp NewCombination

File: app/KtWeb/prgs/Main/www/jkut/libdm/captcha.js
import * as iter from '../_js/iter.js';import * as str from '../_js/str.js';import * as bytes from '../_js/bytes.js';import * as cryp from '../_js/cryp.js';import * as dic from '../_js/dic.js';import * as timer from '../_js/timer.js';import * as js from '../_js/js.js';import * as storage from '../_js/storage.js';import * as sys from '../_js/sys.js';import * as math from '../_js/math.js';import * as domo from '../_js/domo.js';import * as ui from '../_js/ui.js';import * as arr from '../_js/arr.js';import * as time from '../_js/time.js';import * as client from '../_js/client.js';import * as b64 from '../_js/b64.js';




const Q =sys.$checkNull( ui.q);














export  function mk(storeId, counterLimit, zeroColor, oneColor)  {sys.$params(arguments.length, 4);
  const wg =sys.$checkNull( Q("div"));
  const now =sys.$checkNull( time.now());
  const Counter =sys.$checkNull( [getCounter(storeId)]);
  if (sys.asBool(now - getTime(storeId) > 900000)) {
    Counter[0] =sys.$checkExists(Counter[0],sys.$checkNull( 0));
    setCounter(storeId, 0);
    setTime(storeId, time.now());
  }
  const ch0 =sys.$checkNull( arr.fromIter(
    iter.map(iter.$range(0,4), function(i)  {sys.$params(arguments.length, 1);  return Q("input").att("type", "checkbox");})
  ));
  const ch1 =sys.$checkNull( arr.fromIter(
    iter.map(iter.$range(0,4), function(i)  {sys.$params(arguments.length, 1);  return Q("input").att("type", "checkbox");})
  ));

  

  
  
   function check()  {sys.$params(arguments.length, 0);  return sys.asBool(arr.all(ch0, function(ch)  {sys.$params(arguments.length, 1);  return !sys.asBool(ch.isChecked());})) &&
      sys.asBool(arr.all(ch1, function(ch)  {sys.$params(arguments.length, 1);  return ch.isChecked();}))
    ;};


  
  
   function increment()  {sys.$params(arguments.length, 0);
    setCounter(storeId, Counter[0] + 1);
    setTime(storeId, time.now());
  };

  
  
   function isUpLimit()  {sys.$params(arguments.length, 0);  return Counter[0] >= counterLimit;};

  
  
   function reset()  {sys.$params(arguments.length, 0);
    resetCounter(storeId);
    resetTime(storeId);
  };


  

  const tds =sys.$checkNull( arr.cat([
    arr.map(ch0, function(ch)  {sys.$params(arguments.length, 1);  return Q("td")
      .att("style", "border: 1px solid;background-color: " + zeroColor)
      .add(ch)
    ;}),
    arr.map(ch1, function(ch)  {sys.$params(arguments.length, 1);  return Q("td")
      .att("style", "border: 1px solid;background-color: " + oneColor)
      .add(ch)
    ;})
  ]));

  arr.shuffle(tds);
  const tds1 =sys.$checkNull( arr.take(tds, 4));
  const tds2 =sys.$checkNull( arr.drop(tds, 4));

  wg
    .removeAll()
    .add(Q("table")
      .att("border", 0)
      .style("border: 1px solid;background-color: #fffff0")
      .add(Q("tr")
        .adds(tds1))
      .add(Q("tr")
        .adds(tds2)))
  ;

   return {wg:wg, check:check, increment:increment, isUpLimit:isUpLimit, reset:reset};
};


 function getCounter(id)  {sys.$params(arguments.length, 1);
  const N =sys.$checkNull( storage.get(id + "_counter"));
  return sys.asBool( N) ? math.fromStr(N[0])[0] : 0;
};


 function setCounter(id, n)  {sys.$params(arguments.length, 2); storage.put(id + "_counter", math.toStr(n));};


 function resetCounter(id)  {sys.$params(arguments.length, 1); storage.del(id + "_counter");};


 function getTime(id)  {sys.$params(arguments.length, 1);
  const N =sys.$checkNull( storage.get(id + "_time"));
  return sys.asBool( N) ? math.fromStr(N[0])[0] : time.now();
};


 function setTime(id, n)  {sys.$params(arguments.length, 2); storage.put(id + "_time", math.toStr(n));};


 function resetTime(id)  {sys.$params(arguments.length, 1); storage.del(id + "_time");};
